### ISSUE TITLE
Setting for automatically dropping into Ltac debug at failures

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#14321]
This might be a bit tricky to implement, but I would like a setting that automatically drops into debugging mode on any top-level failure, with the frame that's at the bottom of the stack (that is, I want the current frame to be the tactic that first raised the error, not the top-level call that failed) (though this is not essential).  Importantly, this setting should not incur the overhead of `Set Ltac Debug. Set Ltac Batch Debug.`; I don't want my tactic invocation that normally takes 30 seconds to take hours.  I don't want to have to step through the entire tactic to get to this point.

For example, I'd like to do
```coq
Set Ltac Backtrace.
Ltac a := idtac "a"; fail; idtac "a2".
Ltac b := idtac "b"; a; idtac "b2".
Ltac c := idtac "c"; b; idtac "c2".
Goal False.
  Set Ltac Debug On Failure.
  c.
```
and have this be equivalent (in this particular case) to `Set Ltac Debug.` and then typing `r "a"` (for run until the "a").  But this should work even if there's not an `idtac` right before the failure.

cc @jfehrle @herbelin

**Comment by jfehrle** [comment]
I've been thinking about use cases such as yours.  My preliminary thoughts (simplest to hardest):
- Provide a way for the user to view the proof state as of just before the last N backtracking operations.  (The implementation is probably just saving the stack at the right moment plus GUI to make it usable.)
- A similar feature that keeps the last N tactics that were applied.
- Let the user reset the proof to one of those prior states so they can continue execution from that state.
- At breakpoints, let the user to enter tactics in the debugger panel outside of the proof script as a way of experimenting with fixes.  The user would decide whether to include or omit the related proof state changes in the 
proof being debugged.

**Comment by JasonGross** [comment]
The first two are sufficient, as long as I can view things with `Set Printing All` so I can copy/paste the goal and context and play around with it.  Importantly, I'd also like to be able to explore backwards from the failure point, so I can see the proof state before that.

> Let the user reset the proof to one of those prior states so they can continue execution from that state.

This would be useful, though I think you'd need to emit an error at Qed about the proof having been generated via interaction with the debugger.

**Comment by samuelgruetter** [comment]
A related/alternative feature achieving a similar goal would be a setting that causes top-level failures to keep the proof state changes made by the tactic so far, instead of rolling them back.

For example, the following script

```coq
Ltac a := set (A1 := 3); fail; set (A2 := 4).
Ltac b := set (B1 := 2); a; set (B2 := 5).
Ltac c := set (C1 := 1); b; set (C2 := 6).
Goal False.
  Unset Ltac Rollback On Failure.
  c.
```

would result in

```
  C1 := 1 : nat
  B1 := 2 : nat
  A1 := 3 : nat
  ============================
  False
```

If after such a non-rolled-back failure, it would still be possible to enter and execute further Ltac commands in the toplevel proof script, that would be quite a powerful debugging mechanism, without even having to start/use the debugger.

Do you think that would be hard to implement?

**Comment by JasonGross** [comment]
@samuelgruetter I'm not sure how to make technical sense of your request.  If the first branch of a `match` statement fails, should the tactic quit?  If you have `match goal with _ => rewrite a; fail | _ => rewrite b end`, do we get the rewrite from the first branch, or the rewrite from the second?  If I write `try (a; fail)`, should `a` persist?  What about `(a; fail) + idtac`?

**Comment by JasonGross** [comment]
I guess you said "top-level failure", but I don't think that's really sensible concept in the tactic monad.  If you have the moral equivalent of `try (tac1; tac2; fail e) with e -> (tac3; reraise e)`, should `tac1` and `tac2` be rolled back, or not?  What about `try (tac1; tac2; fail e) with e -> reraise e`?  What about `try (tac1; tac2; fail e) with e -> fail e'`?

**Comment by SkySkimmer** [comment]
>Do you think that would be hard to implement?

Basically impossible.

**Comment by samuelgruetter** [comment]
> I guess you said "top-level failure", but I don't think that's really sensible concept in the tactic monad.

You're the one who introduced this concept at the beginning of this thread, @JasonGross :wink: And intuitively, it makes sense to me... What I mean is, if the user processes `my_tac.` in interactive proof mode, and the tactic would fail with an error `e`, make it a success instead, and in the goal buffer, display the proof state just as is was before `e` was (re-)raised the last time.

**Comment by JasonGross** [comment]
@samuelgruetter I guess that's true.  It seems feasible to imagine an approximate implementation of this, something like, "when the overall tactic fails, drop into the debugger and automatically step backwards, with a handful of known constructions that constitute 'propagating the failure'".  I guess you could do the same thing without the debugger?  But it's always going to be a heuristic process, and without the ability to step forwards and backwards, it seems like the heuristic getting it wrong is going to be much more painful.

**Comment by jfehrle** [comment]
I agree with @SkySkimmer: It's unlikely we would want to support such a feature within the tactic monad, but we should be able to do something in the debugger.  For example, the debugger can remember the context and stack for the last successful tactic, which could then be recreated somehow after a top-level failure as a separate proof to allow the user to experiment with alternative tactics.  Users would want to see the call stack and the defined variables from that last tactic.  A richer system would probably save history about the last N failures/backtracking events and/or be smarter about what gets saved to make debugging more efficient.

In the basic form of such a feature, the user would be able to experiment with the context, but wouldn't be able to complete the original failing proof.  If the debugger can efficiently save the partial proof term then it _may_ be possible to permit completing the original proof.

The user would need to be in the debugger to do this, which had a performance cost of about 20-25% the one time I measured it.  Saving history data could use large amounts of memory.  No free lunch, but the cost is likely  affordable.

**Comment by jfehrle** [comment]
> @samuelgruetter A related/alternative feature achieving a similar goal would be a setting that causes top-level failures to keep the proof state changes made by the tactic so far, instead of rolling them back.

Is that flexible enough?  Suppose you have multiple backtracking points and there's a failure in the innermost one that cascdes to a top-level failure.  I think you'd want to be able to use the proof state from the innermost backtracking point, which might require keeping the proof term in the history.

OTOH, maybe supporting this only at the top-level would cover the most common case.  That could be supplemented by a debugger command that lets you interrupt the proof while keeping the proof state changes.

Samuel, how long do your scripts run before they get top-level failures?  How much time would this feature save you?

**Comment by samuelgruetter** [comment]
> Is that flexible enough? Suppose you have multiple backtracking points and there's a failure in the innermost one that cascdes to a top-level failure. 

Are you talking about examples like the following?

```coq
tac1; first [ tac2 | tac3 ]
```

If `tac2` and `tac3` both fail, `first` will fail too, and the displayed error will be a generic `No applicable tactic` produced by `first`, so I'd say the "top-level failure" is the failure of `first`, and the proof state to be displayed should be the one reached after `tac1`.

> Samuel, how long do your scripts run before they get top-level failures? How much time would this feature save you?

My scripts are (currently) quite fast, and I also usually have a clear idea of which part needs debugging, but the feature I'm really missing is being able to interactively try out (and undo) new commands in a proof state that's reached somewhere deep down an Ltac call stack (https://github.com/coq/coq/issues/15066).

My current workaround is [this hack](https://github.com/samuelgruetter/coq/commit/d1a1263d2c4b7bd986af493407ac0893954df62b) that modifies the core definition of the tactic monad (which I think is not the right way of doing it, but it was fun, and might be useful to me right now), so if I know which tactic I want to debug, I can just redefine it as `::= halt`:

```coq
Ltac needs_debugging := fail.

Ltac a := set (A1 := 3); needs_debugging; set (A2 := 4).
Ltac b := set (B1 := 2); a; set (B2 := 5).
Ltac c := set (C1 := 1); b; set (C2 := 6).

Goal False.
  Ltac needs_debugging ::= halt.
  c.
```

results in

```
  C1 := 1 : nat
  B1 := 2 : nat
  A1 := 3 : nat
  ============================
  False
```

so I can now interactively try out various alternative implementations of `needs_debugging`.

**Comment by jfehrle** [comment]
I was thinking about a case more like:

`tac1; first [ tac2 | tac3 ]; long_ltac_tactic; first [ tac4; tac5 ]`

in which the root cause of the failure is in `first [ tac2 | tac3 ]` that leads to an ultimate failure in `first [ tac4; tac5 ]`.
Maybe that's not a realistic (or perhaps even possible) case.  Plenty of time to ponder before I implement anything.

**Comment by jfehrle** [comment]
I have code that preserves the debugger state history (goals, stack, variables) for the last N _successful_ tactics, which you can examine (_not_ re-execute).  The history doesn't currently record when primitive tactics fail, but the user could infer from the history that an Ltac tactic provided that at its first tactic succeeded.  Tracking every failed primitive tactic might generate an excessive amount of history (depending on how the script is written).

Consider:

1.  The debugger could catch a top level failure and let the user examine the history in the debugger starting at the tactic that caused the failure.  (probably simple to implement)  

2.  Provide a way of exiting a failing top-level Ltac tactic without rolling back the proof term.  (a good bit more thought/work)

Gentlemen, WDYT?  How valuable is # 1 without # 2?
